File: python/98.验证二叉搜索树.py
#
# @lc app=leetcode.cn id=98 lang=python3
#
# [98] 验证二叉搜索树
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        callback = []
        traverse_binary_tree(root, callback)

        for i in range(len(callback) - 1):
            if callback[i] > callback[i+1]:
                return False  
        return True

    def traverse_binary_tree(node, callback):
        if node is None:
            return
        traverse_binary_tree(node.left, callback)
        callback.append(node.val)
        traverse_binary_tree(node.right, callback)
# @lc code=end

File: C++/98.验证二叉搜索树.cpp
/*
 * @lc app=leetcode.cn id=98 lang=cpp
 *
 * [98] 验证二叉搜索树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return judgement_function(root, LONG_MIN, LONG_MAX);
    }
    bool judgement_function(TreeNode* root, long long lower, long long upper){
        if (root == nullptr){
            return true;
        }
        if (root->val <= lower || root->val >= upper){
            return false;
        }
        else
            return judgement_function(root->left, lower, root->val) && judgement_function(root->right, root->val, upper);
    }
};

// @lc code=end

